cgp/genome.py

```diff
@@ -267,6 +267,183 @@ def randomize(self, rng: np.random.RandomState) -> None:
         # accept generated dna if it is valid
         self.dna = dna
 
+    def reorder(self, rng: np.random.RandomState) -> None:
+        """Reorder the genome
+
+        Shuffle node ordering of internal (hidden) nodes in genome without changing node behavior.
+        (Goldman 2015, DOI: 10.1109/TEVC.2014.2324539)
+
+        During reordering, inactive genes, e.g., input genes of nodes with arity zero, are not taken
+        into account and can hence have invalid values after reordering.
+        These invalid values are replaced by random values for the respective gene after reordering.
+
+        Parameters
+        ----------
+        rng : numpy.RandomState
+            Random number generator instance.
+
+        Returns
+        ----------
+        None
+        """
+        if not self._n_rows == 1:
+            Warning("n_rows must be 1, reorder is not applied")
+            return
+        if not self._levels_back == self._n_columns:
+            Warning("levels_back must be equal n_columns, reorder is not applied")
+            return
+
+        dna = self._dna.copy()
+
+        node_dependencies = self._determine_node_dependencies()
+
+        addable_nodes = self._get_addable_nodes(node_dependencies)
+
+        node_idx = self._n_inputs  # First position to be placed is after inputs
+        used_node_indices: List[int] = []
+
+        while len(addable_nodes) > 0:
+
+            current_node_idx = rng.choice(list(addable_nodes))
+
+            dna = self._copy_dna_segment(
+                dna, idx_old_location=current_node_idx, idx_new_location=node_idx
+            )
+
+            for dependencies in node_dependencies.values():
+                dependencies.discard(current_node_idx)
+
+            [addable_nodes, used_node_indices] = self._update_addable_nodes(
```

Comment:
```suggestion
            addable_nodes, used_node_indices = self._update_addable_nodes(
```
i think these are not needed